Development docs (#374)
18F/charlie: CONTRIBUTING.md

We want to ensure a welcoming environment for all of our projects. Our staff follow the [18F Code of Conduct](https://github.com/18F/code-of-conduct/blob/master/code-of-conduct.md) and all contributors should do the same.

We encourage you to read this project's CONTRIBUTING policy (you are here), its [LICENSE](LICENSE.md), and its [README](README.md).

If you have any questions or want to read more, check out the [18F Open Source Policy GitHub repository](https://github.com/18f/open-source-policy), or just [shoot us an email](mailto:18f@gsa.gov).

## Development

Charlie is a Slack App, and for structural and policy reasons described below, it can't be fully qualified and exercised in a local development environment.

Because Slack apps work by responding to web hooks rather than setting up a persistent websocket connection, the machine where Charlie runs cannot respond to Slack events unless it is accessible on the public internet over HTTPS. **GSA policy forbids our computers from doing this**. As a result, it is necessary to run the app in cloud.gov to test it.

> To facilitate this requirement, there is a `GSA TTS Testing` Slack instance available to test in-development changes to Charlie before deploying to production. The Slack instance `GSA TTS Testing` is deployed in cloud.gov and this github repository is fully wired up with CI/CD to automatically deploy to it (for branches other than `main`).

Author: When I built the q-expander bot I didn't have to do anything other than `cf push` and wait a few minutes to have that code operational in `GSA TTS Testing` Slack. There has to be some kind of CI/CD glued up, doesn't there? I do think I have something not quite right though...is Slack even involved in any part of the CI/CD bits?

> To facilitate this requirement, there is a `GSA TTS Testing` Slack instance available to test in-development changes to Charlie before deploying to production. The Slack instance `GSA TTS Testing` is deployed in cloud.gov and this github repository is fully wired up with CI/CD to automatically deploy to it (for branches other than `main`).

Reviewer: CI/CD does publish to production on merges to `main`, but it doesn't do anything with the test Slack instance. We could fix that!

> To facilitate this requirement, there is a `GSA TTS Testing` Slack instance available to test in-development changes to Charlie before deploying to production. The Slack instance `GSA TTS Testing` is deployed in cloud.gov and this github repository is fully wired up with CI/CD to automatically deploy to it (for branches other than `main`).

Author: I'm confused...how were my changes on dev branches automatically functional in `GSA TTS Testing` Slack? I don't know of anything I did other than `cf push` to get things working on that Slack.

> To facilitate this requirement, there is a `GSA TTS Testing` Slack instance available to test in-development changes to Charlie before deploying to production. The Slack instance `GSA TTS Testing` is deployed in cloud.gov and this github repository is fully wired up with CI/CD to automatically deploy to it (for branches other than `main`).

Reviewer: I suspect I'm not understanding what you mean, so here's where my head is. When I read this line:

> this github repository is fully wired up with CI/CD to automatically deploy to it

It makes me think that doing a `git push` to my feature branch will result in a new deployment into the `gsa-18f-hubot/dev` space in cloud.gov, which is not the case. You have to run `cf push` from your local machine, which is not automatic.

Temporary Authentication Code ( Get one at https://login.fr.cloud.gov/passcode ):
Authenticating...
OK

Select an org:
1. gsa-18f-hubot
2. sandbox-gsa

Org (enter to skip): 1
Targeted org gsa-18f-hubot.

Select a space:
1. dev
2. prod
3. staging

Space (enter to skip): 1
Targeted space dev.

API endpoint:   https://api.fr.cloud.gov
API version:    3.107.0
user:           <username>@gsa.gov
org:            gsa-18f-hubot
space:          dev
```

At any point you can make sure you are connected to the right org and space with:

    cf target -o gsa-18f-hubot -s dev

Finally, to deploy your development branch to `GSA TTS Testing` Slack, execute this command:

    cf push --vars-file ./dev.yml

Note: it takes a few minutes for the code to fully build and deploy. The console should give updates of its progress and exit to a prompt when finished.

#### Getting the New Bot Deployed to Production

After `awesomobot` has been linted, unit tested, live tested on `GSA TTS Testing` Slack, and deemed production-ready, the process to get the new bot to `TTS` Slack is as follows:

- Push your `awesomo` branch to github.com/18F/Charlie
- File a Pull Request for `awesomo->main`

>   -- PR should be from your development
>   -- Include an explanatory paragraph in the PR as to what's in the request.

Reviewer: ```suggestion
   - PR should be from your development
   - Include an explanatory paragraph in the PR as to what's in the request.
```

Second-level lists in GH Markdown require just indentation. The `--` doesn't trigger list items, so the rendering goes wonky.

Author: Agreed. Believe it or not it's the VS Code auto-formatter that made that change.

> - Request a reviewer. If you don't know who should review the code, ask in the [#bots](https://app.slack.com/client/T025AQGAN/C02FPFGBG) channel

Reviewer: ```suggestion
- Get your PR reviewed. By default @18F/charlie-maintainers will be assigned, and you can, ask in the [#bots](https://app.slack.com/client/T025AQGAN/C02FPFGBG) channel for anyone else to review
```

- After PR has passed review, merge the changes to `main` (sometimes the reviewer does this)

>   -- CI/CD is setup so the deployment to `TTS` Slack is automatic

Reviewer: ```suggestion
   - CI/CD is setup so the deployment to `TTS` Slack is automatic
```
